### PR TITLE
Support for CTAS with no data

### PR DESCRIPTION
**Pull request by kokosing**
Support for CTAS with no data

Add support for queries like:
`CREATE TABLE T AS SELECT ... WITH NO DATA`

This code is based on #3213

Task: #3145

Test Plan: running tests

### CODE REVIEW

**Comment by kokosing** [comment]
CC: @nileema @cberner @zinking I saw you did a review for #3213, can I ask you for a review to this one also?

File: presto-docs/src/main/sphinx/sql/create-table-as.rst [hunk 1]
```diff
@@ -7,7 +7,7 @@ Synopsis
 
 .. code-block:: none
 
-    CREATE TABLE table_name AS query
+    CREATE TABLE table_name AS query [WITH [NO] DATA]
```
Comment (electrum) [inline presto-docs/src/main/sphinx/sql/create-table-as.rst]: Format like this:
```
CREATE TABLE table_name AS query
[ WITH [ NO ] DATA ]

File: presto-docs/src/main/sphinx/sql/create-table-as.rst [hunk 2]
```diff
@@ -24,3 +24,10 @@ Create a new table ``orders_by_date`` that summarizes ``orders``::
     SELECT orderdate, sum(totalprice) AS price
     FROM orders
     GROUP BY orderdate
+
+Create a new table ``empty_nation`` table but with the same schema as ``nation`` but with no data::
+
+    CREATE TABLE empty_nation AS
+    SELECT * 
+    FROM nation WITH NO DATA
```
Comment (electrum) [inline presto-docs/src/main/sphinx/sql/create-table-as.rst]: The `WITH NO DATA` should be at the end of the query, on a new line

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 [hunk 3]
```diff
@@ -69,6 +71,10 @@ with
     : WITH RECURSIVE? namedQuery (',' namedQuery)*
     ;
 
+withOrWithoutData
```
Comment (martint) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: This is used in a single place and it doesn't result in the creation of an independent tree node, so inline it into the rule that uses it.

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 [hunk 4]
```diff
@@ -483,6 +490,7 @@ MAP: 'MAP';
 SET: 'SET';
 RESET: 'RESET';
 SESSION: 'SESSION';
+DATA: 'DATA';
```
Comment (martint) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: Add `DATA` and `NO` to the list of non-reserved tokens.

**Comment by kokosing** [comment]
@martint I applied comments, rebased and squashed. Thanks for comments.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 5]
```diff
@@ -111,6 +112,16 @@ public void setUpdateType(String updateType)
         this.updateType = updateType;
     }
 
+    public boolean getCreateTableAsSelectWithData()
+    {
+        return createTableAsSelectWithData;
+    }
+
+    public void setCreateTableAsSelectWithData(Boolean createTableAsSelectWithData)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: Why `Boolean` instead of `boolean`?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 6]
```diff
@@ -111,6 +112,16 @@ public void setUpdateType(String updateType)
         this.updateType = updateType;
     }
 
+    public boolean getCreateTableAsSelectWithData()
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: This should be named `isCreateTableAsSelectWithData`

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 7]
```diff
@@ -151,7 +151,9 @@ public Node visitUse(SqlBaseParser.UseContext context)
     @Override
     public Node visitCreateTableAsSelect(SqlBaseParser.CreateTableAsSelectContext context)
     {
-        return new CreateTableAsSelect(getQualifiedName(context.qualifiedName()), (Query) visit(context.query()), processTableProperties(context.tableProperties()));
+        boolean withData = context.NO() == null;
+        QualifiedName qualifiedName = getQualifiedName(context.qualifiedName());
```
Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java]: Why introduce a variable for the table name?

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java [hunk 8]
```diff
@@ -711,18 +711,24 @@ public void testCreateTable()
     public void testCreateTableAsSelect()
             throws Exception
     {
+        Query selectStartFromT = simpleQuery(selectList(new AllColumns()), table(QualifiedName.of("t")));
```
Comment (martint) [inline presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java]: I'd name these variables `query` and `table`, respectively.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java [hunk 9]
```diff
@@ -148,43 +132,71 @@ public void testCreateTable()
...
+        assertCreateTableAsSelect(
                 "test_unicode",
                 "SELECT '\u2603' unicode",
                 "SELECT 1");
+
+        assertCreateTableAsSelect(
+                "test_with_data",
+                "SELECT * FROM orders WITH DATA",
+                "SELECT * FROM orders",
+                "SELECT count(*) FROM orders ");
```
Comment (martint) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java]: extra space after `orders`

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java [hunk 10]
```diff
@@ -730,7 +736,7 @@ public void testCreateTableAsSelect()
                                         new StringLiteral("ban"),
                                         new StringLiteral("ana"))))
                                 .put("a", new ArrayConstructor(ImmutableList.of(new StringLiteral("v1"), new StringLiteral("v2"))))
-                                .build()));
+                                .build(), true));
```
Comment (martint) [inline presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java]: `true` should be on the line below

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java [hunk 11]
```diff
@@ -140,9 +141,15 @@ private RelationPlan createTableWriterPlan(Analysis analysis, RelationPlan plan,
                 symbolAllocator.newSymbol("partialrows", BIGINT),
                 symbolAllocator.newSymbol("fragment", VARBINARY));
 
-        TableWriterNode writerNode = new TableWriterNode(
+        PlanNode source = plan.getRoot();
+
+        if (! analysis.getCreateTableAsSelectWithData()) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: No space after `!`

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java [hunk 12]
```diff
@@ -55,22 +55,6 @@ protected AbstractTestDistributedQueries(QueryRunner queryRunner, Session sample
         super(queryRunner, sampledSession);
     }
 
-    private void assertCreateTable(String table, @Language("SQL") String query, @Language("SQL") String rowCountQuery)
```
Comment (martint) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java]: Not a big deal, but try to avoid renaming and moving methods in the same commit (especially since the move is unrelated to this change), as it makes it harder to see if there are meaningful changes. What I usually do in these cases is introduce another commit before or after to do some "cleanup".

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java (hunk 7 above)
Comment (kokosing) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java]: I wanted to make to make a return statement to have less code, because there was happening a lot in just a single line.

In-lined.

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 (hunk 4 above)
Comment (kokosing) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: done

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java (hunk 12 above)
Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java]: Ok, thanks. Do you do separate PR for that commit?

**Comment by kokosing** [comment]
@martint Thanks for comments. I have just addressed them all and rebased.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java (hunk 12 above)
Comment (electrum) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java]: Another commit in same PR is fine

**Comment by kokosing** [comment]
Travis build failed, but it is passing on my sandbox.

**Comment by martint** [comment]
Merged, thanks!
